README.md
# PHP-LazyList

A lazy-list (or generator, or fused-stream, or seq, or whatever) library for PHP.

Supports PHP >=7.4.

## Usage

### Creation 

Lazy lists can be created from arrays
```php
$lazyList = \Spencerwi\Lazy_list\LazyList::fromArray([1,2,3,4,5]);
```

They can also be created from callables that take an `int` index and return 
 either some value or else `\Spencerwi\Lazy_list\LazyList::STOP_ITERATION` to 
 signal the end of the list.

```php
$lazyList2 = new \Spencerwi\Lazy_list\LazyList(function(int $i) {
    if ($i < 10) {
        return $i;
    } else {
        return \Spencerwi\Lazy_list\LazyList::STOP_ITERATION;
    }
});
// This amounts to a LazyList containing [0,1,2,3,4,5,6,7,8,9].
```

### `map`

You can map over a lazy list, and it's a "lazy" map, so it doesn't call
your mapper function until you actually force evaluation of the list.

```php
$mapperWasCalled = false;
$squares = $lazyList->map(function($i): int use (&$mapperWasCalled) {
    $mapperWasCalled = true;
    return $i * $i;
});
// $mapperWasCalled is still false!
```

### `filter`

You can apply a filter predicate to a lazy list, and it'll "lazily" apply the 
filter, meaning that your filter function doesn't get called until you actually
force evaluation of the list.

```php
$filterFnWasCalled = false;
$oddNumbers = $lazyList->filter(function (int $i): bool use (&$filterFnWasCalled) {
    $filterFnWasCalled = true;
    return ($i % 2 === 1);
});
// $filterFnWasCalled is still false!
```

### Iteration

You can also iterate over a lazy list more traditionally; it's an iterator.

```php
foreach ($squares as $square) {
    echo $square . "\n";
}
/* Output:
 * 1
 * 4
 * 9
 * 16
 * 25
 */
```

It even supports index => value iteration:

```php
foreach ($squares as $index => $square) {
    echo $index . ": " . $square . "\n";
}
/* Output:
 * 0: 1
 * 1: 4
 * 2: 9
 * 3: 16
 * 4: 25
 */
```

### `take($count)`

You can also take just a certain number of elements from the beginning:

```php
$l = \Spencerwi\Lazy_list\LazyList::fromArray([1,2,3,4,5]);
$l->take(2); // returns the array [1,2]

// What happens when we take too many?
$l->take(99); // we get the whole list as an array: [1,2,3,4,5]
```

### `toArray()`

You can more directly dump the whole list out to an array with `toArray()`:

```php
$l = \Spencerwi\Lazy_list\LazyList::fromArray([1,2,3,4,5]);
$->toArray(); // [1,2,3,4,5]

```

**BE VERY CAREFUL WITH THIS**, as you can wind up using a generator to create an
infinite list (if your generator never returns null), which *will* cause an 
infinite loop if you call toArray() on it. Unfortunately, there's no way to know
ahead of time that a generator is infinite, or else we'd have the list throw an 
exception on trying to toArray() an infinite LazyList.

### `reduce()`

Alongside `map`, you also have its friend `reduce`, which takes an initial value,
and a function that works "pairwise" along the list calling a function you 
provide, first on initial and the first element, then on the previous result
and the second element, then on _that_ result and the third element, and so on.

```php
$l = \Spencerwi\Lazy_list\LazyList::fromArray([2,3,4]);
$sum = $l->reduce(1, function(int $previous, int $current) {
  return $previous + $current;
});
// $sum is now ((1+2)+3)+4, that is, 10. 
```

If you try to "reduce" an empty list, you just get the "initial" value back. 
That's why it's there.

```php
$l = \Spencerwi\Lazy_list\LazyList::fromArray([]);
$sum = $l->reduce(1, function(int $previous, int $current) {
  return $previous + $current;
});
// $sum is now just 1, since the list was empty
```

**THIS OPERATION IS "EAGER"**, meaning that **it evaluates the whole list**. As with
`toArray()`, that means that **if you have an infinite LazyList, `reduce()` will 
cause an infinite loop.**



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added 
 - Everything! 



tests/LazyListTest.php
<?php
declare(strict_types=1);

use PHPUnit\Framework\TestCase;
use Spencerwi\Lazy_list\LazyList;

final class LazyListTest extends TestCase {

    public function test_canBeCreatedFromAnArray(): void 
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = LazyList::fromArray([1,2,3]);

        $this->assertEquals(
            [1,2,3],
            $lazyList->toArray()
        );
    }

    public function test_canBeCreatedFromAGenerator(): void 
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = new LazyList(function (int $i) {
            if ($i <= 5) {
                return $i * $i;
            } else {
                return LazyList::STOP_ITERATION;
            }
        });

        $this->assertEquals(
            [0, 1, 4, 9, 16, 25],
            $lazyList->toArray()
        );
    }

    public function test_map_FromArray(): void
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = LazyList::fromArray([1,2,3]);

        $lazySquares = $lazyList->map(function(int $x) { 
            return $x * $x; 
        });

        $this->assertEquals(
            [1, 4, 9],
            $lazySquares->toArray()
        );
    }

    public function test_map_FromGenerator(): void
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = new LazyList(function(int $i) {
            if ($i < 4) {
                return $i;
			}
			return LazyList::STOP_ITERATION;
        });

        $lazySquares = $lazyList->map(function(int $x) { 
            return $x * $x; 
        });

        $this->assertEquals(
            [0, 1, 4, 9],
            $lazySquares->toArray()
        );
    }

    public function test_map_IsLazy(): void
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = LazyList::fromArray([1,2,3]);

        $mapperFunctionWasInvoked = false;
        $lazyList->map(function(int $i) use (&$mapperFunctionWasInvoked) {
            $mapperFunctionWasInvoked = true;
        });

        $this->assertFalse($mapperFunctionWasInvoked);
    }

    public function test_reduce_FromArray(): void
    {
        /** @var LazyList $lazyList */
        $lazyList = LazyList::fromArray([2,3,4]);

        /** @var int $result */
        $result = $lazyList->reduce(1, fn(int $previous, int $current): int => $previous + $current);

        $this->assertEquals(1+2+3+4, $result);
    }

    public function test_reduce_FromEmptyArray(): void
    {
        /** @var LazyList $lazyList */
        $lazyList = LazyList::fromArray([]);

        /** @var int $result */
        $result = $lazyList->reduce(1, fn(int $previous, int $current): int => $previous + $current);

        $this->assertEquals(1, $result);
    }

    public function test_reduce_FromGenerator(): void 
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = new LazyList(function(int $i) {
            if ($i < 4) {
                return $i + 1;
            } else {
                return LazyList::STOP_ITERATION;
            }
        });

        /** @var int $result */
        $result = $lazyList->reduce(1, fn(int $previous, int $current): int => $previous + $current);

        $this->assertEquals(1+1+2+3+4, $result);
    }

    public function test_reduce_FromEmptyGenerator(): void
    {
        /** @psalm-var LazyList<string> $lazyList */
        $lazyList = new LazyList(fn(int $i): string => LazyList::STOP_ITERATION);

        /** @var string $result */
        $result = $lazyList->reduce('', fn(string $previous, string $current): string => $previous . $current);

        $this->assertEquals('', $result);
    }

	public function test_filter_FromArray(): void
	{
        /** @psalm-var LazyList<string> $lazyList */
        $lazyList = LazyList::fromArray(['keep', 'drop', 'keep', 'keep', 'drop']);

		$keeps = $lazyList->filter(function(string $element) {
			return ($element === 'keep');
		});

		$this->assertEquals(
			['keep', 'keep', 'keep'],
			$keeps->toArray()
		);
	}

	public function test_filter_FromGenerator(): void
	{
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = new LazyList(function(int $i){
            if ($i < 4) {
                return $i + 1;
            }
            return LazyList::STOP_ITERATION;
        });

		$oddLazies = $lazyList->filter(fn(int $element): bool => ($element % 2 === 1));

		$this->assertEquals(
			[1, 3],
			$oddLazies->toArray()
		);
	}

    public function test_filter_IsLazy(): void
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = new LazyList(function(int $i){
            if ($i < 4) {
                return $i + 1;
            }
            return LazyList::STOP_ITERATION;
        });

        $filterFnWasCalled = false;

		$oddLazies = $lazyList->filter(function(int $element) use (&$filterFnWasCalled) {
            $filterFnWasCalled = true;
			return ($element % 2 === 1);
		});

		$this->assertFalse($filterFnWasCalled, 'The filter predicate should not be called until the LazyList is eager-evaluated!');
    }

    public function test_filter_then_map_FromArray(): void
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = LazyList::fromArray([0,1,2,3,4,5,6,7,8,9]);

		$oddLaziesSquared = $lazyList->filter(fn(int $element): bool => ($element % 2) === 1)
		    ->map(fn(int $element): int => $element * $element);

        $this->assertEquals(
            [1, 9, 25, 49, 81],
            $oddLaziesSquared->toArray()
        );
    }

    public function test_filter_then_map_FromGenerator(): void
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = new LazyList(function(int $i){
            if ($i < 10) {
                return $i + 1;
            }
            return LazyList::STOP_ITERATION;
        });

		$oddLaziesSquared = $lazyList->filter(fn(int $element): bool => ($element % 2) === 1)
		    ->map(fn(int $element): int => $element * $element);

        $this->assertEquals(
            [1, 9, 25, 49, 81],
            $oddLaziesSquared->toArray()
        );
    }


    public function test_map_then_filter_FromArray(): void
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = LazyList::fromArray([0,1,2,3,4,5,6,7,8,9]);

		$oddLaziesSquared = $lazyList->map(fn(int $element): int => $element * $element)
            ->filter(fn(int $element): bool => ($element % 2) === 1);

        $this->assertEquals(
            [1, 9, 25, 49, 81],
            $oddLaziesSquared->toArray()
        );
    }

    public function test_map_then_filter_FromGenerator(): void
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = new LazyList(function(int $i){
            if ($i < 10) {
                return $i + 1;
            }
            return LazyList::STOP_ITERATION;
        });

		$oddLaziesSquared = $lazyList->map(fn(int $element): int => $element * $element)
            ->filter(fn(int $element): bool => ($element % 2) === 1);

        $this->assertEquals(
            [1, 9, 25, 49, 81],
            $oddLaziesSquared->toArray()
        );
    }

    public function test_CanBeIterated_FromArray(): void
    {
        $originalArray = [1,2,3];
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = LazyList::fromArray($originalArray);

        $i = 0;
        foreach ($lazyList as $index => $element) {
            $this->assertEquals(
                $i, 
                $index,
                "The iteration key should be the current index!"
            );
            $this->assertEquals(
                $originalArray[$index],
                $element,
                "The list should be iterated correctly!"
            );
            $i++;
        }
    }

    public function test_CanBeIterated_FromGenerator(): void
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = new LazyList(function (int $i) {
            if ($i < 4) {
                return $i + 1;
            } 
			return LazyList::STOP_ITERATION;
        });

        $i = 0;
        foreach ($lazyList as $index => $element) {
            $this->assertEquals(
                $i, 
                $index,
                "The iteration key should be the current index!"
            );
            $this->assertEquals(
                $i + 1,
                $element,
                "The list should be iterated correctly!"
            );
            $i++;
        }
    }

    public function test_filter_then_iterate_FromArray(): void
    {
        $originalArray = [1,2,3,4,5,6,7,8,9];
        $expectedResultOrder = [1,3,5,7,9];
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = LazyList::fromArray($originalArray);

        $oddLazies = $lazyList->filter(fn(int $element): bool => ($element % 2) === 1);

        $i = 0;
        $seen = [];
        foreach ($oddLazies as $index => $element) {
            $this->assertEquals(
                $i, 
                $index,
                "The iteration key should be the current index!"
            );
            $seen[] = $element;
            $i++;
        }
        $this->assertEquals($expectedResultOrder, $seen, 'Iteration order did not match expected');
    }

    public function test_filter_then_iterate_FromGenerator(): void
    {
        /** @psalm-var LazyList<int> $lazyList */
        $lazyList = new LazyList(function (int $i) {
            if ($i < 10) {
                return $i + 1;
            } 
			return LazyList::STOP_ITERATION;
        });
        $expectedResultOrder = [1,3,5,7,9];

        $oddLazies = $lazyList->filter(fn(int $element): bool => ($element % 2) === 1);

        $i = 0;
        $seen = [];
        foreach ($oddLazies as $index => $element) {
            $this->assertEquals(
                $i, 
                $index,
                "The iteration key should be the current index!"
            );
            $seen[] = $element;
            $i++;
        }
        $this->assertEquals($expectedResultOrder, $seen, 'Iteration order did not match expected');
    }


    /**
     * @dataProvider data_test_take_FromArray
     */
    public function test_take_FromArray(array $inputArray, int $count, array $expectedResult): void
    {
        /** @var LazyList $lazyList */
        $lazyList = LazyList::fromArray($inputArray);

        $this->assertEquals(
            $expectedResult,
            $lazyList->take($count)
        );
    }
    /**
     * DataProvider for @see test_take_FromArray
     * @return array test cases in the form [inputArray, count, expectedResult]
     */
    public static function data_test_take_FromArray(): array 
    {
        return [
            "Take some but not all" => [[1,2,3,4,5], 3, [1,2,3]],
            "Take all" => [[1,2,3,4,5], 5, [1,2,3,4,5]],
            "Take none" => [[1,2,3,4,5], 0, []],
            "Take more than all" => [[1,2,3,4,5], 99, [1,2,3,4,5]],
        ];
    }

    /**
     * @dataProvider data_test_filter_then_take_FromArray
     */
    public function test_filter_then_take_FromArray(array $inputArray, callable $predicate, int $count, array $expectedResult): void
    {
        /** @var LazyList $lazyList */
        $lazyList = LazyList::fromArray($inputArray);

        $this->assertEquals(
            $expectedResult,
            $lazyList->filter($predicate)->take($count)
        );
    }
    /**
     * DataProvider for @see test_filter_then_take_FromArray
     * @return array test cases in the form [inputArray, predicate, count, expectedResult]
     */
    public static function data_test_filter_then_take_FromArray(): array 
    {
        $isOdd = fn(int $i): bool => ($i % 2) === 1;
        return [
            "Take some of resulting list but not all" => [[1,2,3,4,5,6,7,8,9], $isOdd, 3, [1,3,5]],
            "Take all of resulting list" => [[1,2,3,4,5,6,7,8,9], $isOdd, 5, [1,3,5,7,9]],
            "Take none" => [[1,2,3,4,5], $isOdd, 0, []],
            "Take more than all" => [[1,2,3,4,5], $isOdd, 99, [1,3,5]],
        ];
    }

    /**
     * @dataProvider data_test_take_FromGenerator
     */
    public function test_take_FromGenerator(callable $generator, int $count, array $expectedResult): void
    {
        /** @var LazyList $lazyList */
        $lazyList = new LazyList($generator);

        $this->assertEquals(
            $expectedResult,
            $lazyList->take($count)
        );
    }
    /**
     * DataProvider for @see test_take_FromGenerator
     * @return array test cases in the form [generator, count, expectedResult]
     */
    public static function data_test_take_FromGenerator(): array 
    {
        return [
            "Take some but not all" => [static::generatorUpTo(4), 3, [0,1,2]],
            "Take all" => [static::generatorUpTo(4), 5, [0,1,2,3,4]],
            "Take none" => [static::generatorUpTo(4), 0, []],
            "Take more than all" => [static::generatorUpTo(4), 99, [0,1,2,3,4]],
        ];
    }


    /**
     * @dataProvider data_test_filter_then_take_FromGenerator
     */
    public function test_filter_then_take_FromGenerator(callable $generator, callable $predicate, int $count, array $expectedResult): void
    {
        /** @var LazyList $lazyList */
        $lazyList = new LazyList($generator);

        $this->assertEquals(
            $expectedResult,
            $lazyList->filter($predicate)->take($count)
        );
    }
    /**
     * DataProvider for @see test_filter_then_take_FromGenerator
     * @return array test cases in the form [generator, predicate, count, expectedResult]
     */
    public static function data_test_filter_then_take_FromGenerator(): array 
    {
        $isOdd = fn(int $i): bool => ($i % 2) === 1;
        return [
            "Take some of resulting list but not all" => [static::generatorUpTo(9), $isOdd, 3, [1,3,5]],
            "Take all of resulting list" => [static::generatorUpTo(9), $isOdd, 5, [1,3,5,7,9]],
            "Take none" => [static::generatorUpTo(5), $isOdd, 0, []],
            "Take more than all" => [static::generatorUpTo(5), $isOdd, 99, [1,3,5]],
        ];
    }

    /**
     * Convenience method for creating a generator that returns 0...$max
     * @return callable a generator function.
     */
    private static function generatorUpTo(int $max) : callable 
    {
        return function(int $i) use ($max) {
            if ($i <= $max)  {
                return $i;
            } else {
                return LazyList::STOP_ITERATION;
            }
        };

    }
}


src/LazyList.php
<?php

namespace Spencerwi\Lazy_list;

/**
 * A Lazy List implementation that can be created either from
 * an array ({@see LazyList::fromArray}) or from a "generator"
 * callable that accepts the current index and returns either a value
 * or else {@see LazyList::STOP_ITERATION} to indicate the end of the list.
 * 
 * @template T
 */
class LazyList implements \Iterator {
	public const STOP_ITERATION = "\Spencerwi\Lazy_list\LazyList::STOP_ITERATION_TOKEN";
    private const SKIP_ELEMENT = "\Spencerwi\Lazy_list\LazyList::SKIP_ELEMENT_TOKEN";

    /** 
     * @var callable $generatorFn 
     * @psalm-var callable(int):T|self::STOP_ITERATION|self::SKIP_ELEMENT
     */
    private $generatorFn;
    
    /** @var int $currentIndex */
    private $currentIndex;

    /** @var int $skipOffset used for iterating with a "skip offset" so that we iterate consistently without repeating even after a filter is applied */
    private $skipOffset = 0;

    /**
     * Constructs a new LazyList from a "generator function" that takes the 
     * current index as an int and returns either a value or {@see LazyList::STOP_ITERATION}.
	 *
     * For example:
     *
     *   $lazyList = new LazyList(new function(int $i) {
     *     if ($i <= 5) {
     *       return $i * $i;
     *     } else {
     *       return LazyList::STOP_ITERATION;
     *     }
     *   });
     *
     * resulting in a LazyList of [0, 1, 4, 9, 16, 25];
     *
     * @param callable $generatorFn 
     * @psalm-param callable(int):T|self::STOP_ITERATION|self::SKIP_ELEMENT $generatorFn
     */
    public function __construct(callable $generatorFn) 
    {
        $this->generatorFn = $generatorFn;
        $this->currentIndex = 0;
    }

    /**
     * Creates a lazy wrapper around an existing array.
     *
     * @template TArrayElement
     * @psalm-param array<TArrayElement> $arr
     * @psalm-return LazyList<TArrayElement> a lazy wrapper around the array.
     * @return LazyList a lazy wrapper around the array.
     */
    public static function fromArray(array $arr) : LazyList
    {
        return new LazyList(
            function (int $i) use ($arr) {
				if ($i >= count($arr)) {
					return static::STOP_ITERATION;
				}
                return $arr[$i];
            }
        );
    }

    /** 
     * Returns a new LazyList that will, on evaluation, apply the given function
     * to every element.
     *
     * @template R
     * @psalm-param callable(T):R|self::STOP_ITERATION|self::SKIP_ELEMENT $f the function to apply to every element in the new LazyList.
     * @param callable $f the function to apply to every element in the new LazyList.
     * @psalm-return LazyList<R> a new LazyList with $f "lazily applied" to it.
     * @return LazyList
     */
    public function map(callable $f) : LazyList
    {
		// I need a variable so I can hand this callable property into the 
		// closure.
		$generatorFn = $this->generatorFn;
        return new LazyList(
            function (int $i) use ($f, $generatorFn) {
				// Do an explicit check for the end-of-list to ensure we don't 
				// over-iterate forever and so that callers don't have to 
				// manually do so.
				$next = $generatorFn($i);
                if ($this->shouldSkip($next)) {
                    return static::SKIP_ELEMENT;
                }
				if (!$this->isStopElement($next)) {
					return $f($next);       
				}
				return static::STOP_ITERATION;
            }
        );
    }

    /**
     * Lazily-applies a {@see callable} filter function to this list, so that 
     *  when you map/iterate/toArray/reduce later, the filter function will be
     *  used to skip elements that don't match your given predicate. Your 
     *  predicate should accept an element of the type in this list, and return
     *  a truthy or falsy value.
     * @param callable $predicate 
     * @psalm-param callable(T):bool $predicate 
     * @psalm-return LazyList<T>
     * @return LazyList
     */
	public function filter(callable $predicate): LazyList
	{
		$generatorFn = $this->generatorFn;
		return new LazyList(
			function (int $i) use ($predicate, $generatorFn) {
                $next = $generatorFn($i);
                if ($this->isStopElement($next)) {
                    return static::STOP_ITERATION;
                }

                $predicateIsSatisfied = $predicate($next);
                if ($predicateIsSatisfied) {
                    return $next;
                }
                return static::SKIP_ELEMENT;
			}
		);
	}

    /**
     * Reduces a LazyList into a single value, immediately evaluating the list.
     * The $reducerFn should accept (previous, current), starting with $initial 
     * as the first "previous" value, and return a value that will either be 
     * the next "previous" or the final result (if there's nothing left in the
     * list). For example:
     *
     *  $lazyList = LazyList::fromArray([1,2,3,4]);
     *  $result = $lazyList->reduce(0, function($total, $current){
     *     return $total + $current;
     *  });
     *  // $result is now 0+1+2+3+4, which is 10
     * 
     * @template TResult
     * @param mixed $initial the "initial" value to use
     * @psalm-param TResult $initial the "initial" value to use
     * @param callable $reducerFn
     * @psalm-param callable(TResult,T):TResult $reducerFn
     * @return mixed
     * @psalm-return TResult
     **/
    public function reduce($initial, callable $reducerFn)
    {
        $result = $initial;
		$i = 0;
        $next = $this->getAtInternal($i);
        while (!$this->isStopElement($next)) {
            if (!$this->shouldSkip($next)) {
                /** @psalm-var T $next */
                $result = $reducerFn($result, $next);
            }
            $i++;
            $next = $this->getAtInternal($i);
        }
        return $result;
    }

	/**
	 * Dumps the entire list out to an array. BEWARE: infinite lists will 
	 * loop infinitely here!
     * @psalm-return array<T>
     * @return array
	 */
    public function toArray() : array 
    {
        $arr = [];
        $i = 0;
        $next = $this->getAtInternal($i);
        while (!$this->isStopElement($next)) {
            if (!$this->shouldSkip($next)) {
                /** @psalm-var T $next */
                $arr[] = $next;
            }
            $i++;
            $next = $this->getAtInternal($i);
        }
        return $arr;
    }

	/**
	 * Gets a subarray of this list, taking the first $count values.
	 * If $count exceeds the length of this list, the entire list is returned,
	 * without "padding" nulls.
	 *
	 * @param int $count how many elements to take.
	 * @psalm-return array<T> the first $count values in this list.
	 * @return array the first $count values in this list.
	 */
    public function take(int $count) : array 
    {
        if ($count > 0) {
            $i = 0;
            $countTaken = 0;
            $arr = [];
            $next = $this->getAtInternal($i);
            while (!$this->isStopElement($next) && $countTaken < $count) {
                if (!$this->shouldSkip($next)) {
                    /** @psalm-var T $next */
                    $arr[] = $next;
                    $countTaken++;
                }
                $i++;
                $next = $this->getAtInternal($i);
            }
			return $arr;
        } else {
            return [];
        }
    }

    /**
	 * Internal getter for a value for a specific index from this LazyList, that 
     *  can return "special" values like {@see self::STOP_ITERATION} or {@see self::SKIP_ELEMENT}
	 * @param int $index
     * @psalm-return T|self::STOP_ITERATION|self::SKIP_ELEMENT
	 * @return mixed the value at that index, or the token that's at that index
	 */
    private function getAtInternal(int $index)
    {
		// PHP is weird about letting you call member properties that are
		// callables.
		$generatorFn = $this->generatorFn; 
		return $generatorFn($index);
    }

	/**
	 * Returns the value at the current spot in the list.
	 *
     * @psalm-return T
	 * @return mixed
	 */
	public function current() 
	{
        do {
            $result = $this->getAtInternal($this->currentIndex + $this->skipOffset);
            if ($this->shouldSkip($result)) {
                $this->skipOffset += 1; // increment our "skip offset" so we remember that we're offset by one.
            }
        } while ($this->shouldSkip($result));
        /** @psalm-var T $result at this point, we've confirmed that it's not a SKIP_ELEMENT, and iterator magic should protect us from STOP_ITERATION */
		return $result;
	}

	/**
	 * Returns the current iteration index in this list.
	 *
	 * @return int the current iteration index.
	 */
	public function key() : int 
	{
		return $this->currentIndex;
	}

    /** 
     * Advances one spot in the list.
     */
    public function next() 
    {
        $this->currentIndex++;
    }

    /**
     * Rewinds the iterator "cursor" to the beginning of the list.
     */
    public function rewind() 
    {
        $this->currentIndex = 0;
        $this->skipOffset = 0;
    }

	/** 
	 * Determines if the current location is "valid" for iteration,
	 * i.e. if the current iteration index has a value. This requires
	 * invoking the generator function, so be aware of that.
	 */
	public function valid() : bool
	{
		return ($this->current() !== static::STOP_ITERATION);
	}


    /**
     * @psalm-param T|self::SKIP_ELEMENT|self::STOP_ITERATION $element 
     * @psalm-assert-if-false T|self::SKIP_ELEMENT $element
     * @psalm-assert-if-true self::STOP_ITERATION $element
     * @return bool 
     */
    private function isStopElement($element): bool
    {
        return ($element === static::STOP_ITERATION);
    }

    /**
     * @psalm-param T|self::SKIP_ELEMENT|self::STOP_ITERATION $element 
     * @psalm-assert-if-false T|self::STOP_ITERATION $element
     * @psalm-assert-if-true self::SKIP_ELEMENT $element
     * @return bool 
     */
    private function shouldSkip($element): bool
    {
        return ($element === static::SKIP_ELEMENT);
    }
}
